### ISSUE TITLE
Julia 1.10.0 segfault at test exit (Windows-only)

### ISSUE THREAD

**Issue by timholy** [timholy/Revise.jl#789]
Even if I comment out the execution of `non_jl_test.jl`, after the sigtest I get a segfault during execution of `_atexit`:
```julia
beginning signatures tests
WARNING: Method definition getindex(Array{T, N} where N where T, Int64) in module Base at essentials.jl:13 overwritten at C:\Users\timho\.julia\juliaup\julia-1.10.0+0.x64.w64.mingw32\share\julia\base\essentials.jl:13.
WARNING: Method definition getindex(Array{T, N} where N where T, Int64, Int64, Int64...) in module Base at essentials.jl:14 overwritten at C:\Users\timho\.julia\juliaup\julia-1.10.0+0.x64.w64.mingw32\share\julia\base\essentials.jl:14.
WARNING: redefinition of constant Base.env_dict. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant Base.env_lock. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant LineEdit.default_console_mode_ref. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant LineEdit.default_console_mode_assigned. This may fail, cause incorrect answers, or produce other errors.
 35.640177 seconds (62.17 M allocations: 3.729 GiB, 3.18% gc time, 45.96% compilation time: 71% of which was recompilation)
Test Summary:   | Pass
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ff98114d753 -- jl_egal_ at C:/workdir/src\julia.h:1512 [inlined]
jl_table_peek_bp at C:/workdir/src\iddict.c:129 [inlined]
ijl_eqtable_get at C:/workdir/src\iddict.c:158
in expression starting at C:\Users\timho\.julia\dev\Revise\test\runtests.jl:3776
jl_egal_ at C:/workdir/src\julia.h:1512 [inlined]
jl_table_peek_bp at C:/workdir/src\iddict.c:129 [inlined]
ijl_eqtable_get at C:/workdir/src\iddict.c:158
get at .\iddict.jl:102 [inlined]
memoized_env_lookup at .\env.jl:14
access_env at .\env.jl:28
get at .\env.jl:156 [inlined]
repl_color at .\client.jl:16
info_color at .\client.jl:24 [inlined]
print_test_results at C:\workdir\usr\share\julia\stdlib\v1.10\Test\src\Test.jl:1155
print_test_results at C:\workdir\usr\share\julia\stdlib\v1.10\Test\src\Test.jl:1117 [inlined]
#finish#43 at C:\workdir\usr\share\julia\stdlib\v1.10\Test\src\Test.jl:1188
finish at C:\workdir\usr\share\julia\stdlib\v1.10\Test\src\Test.jl:1170
unknown function (ip: 0000028dcb28d20b)
jl_apply at C:/workdir/src\julia.h:1982 [inlined]
do_call at C:/workdir/src\interpreter.c:126
eval_value at C:/workdir/src\interpreter.c:223
eval_body at C:/workdir/src\interpreter.c:489
jl_interpret_toplevel_thunk at C:/workdir/src\interpreter.c:775
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:934
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:877
ijl_toplevel_eval at C:/workdir/src\toplevel.c:943 [inlined]
ijl_toplevel_eval_in at C:/workdir/src\toplevel.c:985
eval at .\boot.jl:385 [inlined]
include_string at .\loading.jl:2070
_include at .\loading.jl:2130
include at .\client.jl:489
unknown function (ip: 0000028d8d61016b)
jl_apply at C:/workdir/src\julia.h:1982 [inlined]
do_call at C:/workdir/src\interpreter.c:126
eval_value at C:/workdir/src\interpreter.c:223
eval_stmt_value at C:/workdir/src\interpreter.c:174 [inlined]
eval_body at C:/workdir/src\interpreter.c:635
jl_interpret_toplevel_thunk at C:/workdir/src\interpreter.c:775
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:934
jl_toplevel_eval_flex at C:/workdir/src\toplevel.c:877
ijl_toplevel_eval at C:/workdir/src\toplevel.c:943 [inlined]
ijl_toplevel_eval_in at C:/workdir/src\toplevel.c:985
eval at .\boot.jl:385 [inlined]
exec_options at .\client.jl:291
_start at .\client.jl:552
jfptr__start_86719.1 at C:\Users\timho\.julia\juliaup\julia-1.10.0+0.x64.w64.mingw32\lib\julia\sys.dll (unknown line)
jl_apply at C:/workdir/src\julia.h:1982 [inlined]
true_main at C:/workdir/src\jlapi.c:582
jl_repl_entrypoint at C:/workdir/src\jlapi.c:731
mainCRTStartup at C:/workdir/cli\loader_exe.c:58
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 67158120 (Pool: 67122021; Big: 36099); GC: 96
```
The failure is pretty consistently in an `iddict.c` operation, typically (or perhaps always) https://github.com/JuliaLang/julia/blob/acddc62c1acca478d39c73f8c248d519ed4cfe93/src/iddict.c#L158. It's always possible that Revise corrupts something, as it selectively re-evaluates enough of Base to extract signatures and source locations for virtually all methods. But this doesn't happen on Julia 1.10 on other platforms, nor on Windows with Julia 1.6 or 1.9.

EDIT: can be triggered with just
```julia
julia> Pkg.test(; test_args=["Base signatures"])
```
from within the Revise environment.

**Comment by BioTurboNick** [comment]
The iddict `env_dict` being read from is a const global only present on Windows.

So it's looking up a hash in the iddict, and it is not NULL, but the object it finds there is corrupted such that it cannot be dereferenced to obtain the `header` field...

There's a lock on writes to `env_dict`, suggesting concurrent accesses are expected. Does the read operation also need to be locked to ensure it won't read a value while it is being written to?

**Comment by Zentrik** [comment]
I don't see a segfault on 14 day old nightly on Windows 64 bit. I do see a test failure though,
```
Precompiling project...
  ✓ IndirectArrays
  ✓ CustomUnitRanges
  ✓ UnsafeArrays
  ✓ Example
  ✓ EndpointRanges
  ✓ EponymTuples
  ✓ RoundingIntegers
  ✓ MappedArrays
  ✓ CodeTracking
  ✓ OffsetArrays
  ✓ CatIndices
  ✓ MacroTools
  ✓ JuliaInterpreter
  ✓ LoweredCodeUtils
  ✓ Revise
  15 dependencies successfully precompiled in 29 seconds. 30 already precompiled.
  1 dependency had output during precompilation:
┌ Revise
│  ┌ Warning: precompile directive
│  │     precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, typeof(pairs(NamedTuple())), typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
│  │ failed. Please report an issue in Revise (after checking for duplicates) or remove this directive.
│  └ @ Revise ~/.julia/dev/Revise/src/precompile.jl:65
│  ┌ Warning: precompile directive
│  │     precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:skip_include,), Tuple{Bool}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
│  │ failed. Please report an issue in Revise (after checking for duplicates) or remove this directive.
│  └ @ Revise ~/.julia/dev/Revise/src/precompile.jl:66
│  ┌ Warning: precompile directive
│  │     precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, typeof(methods_by_execution!), Any, MI, DocExprs, Frame, Vector{Bool}})
│  │ failed. Please report an issue in Revise (after checking for duplicates) or remove this directive.
│  └ @ Revise ~/.julia/dev/Revise/src/precompile.jl:69
└  
     Testing Running tests...
Test Summary: |Time
Revise        | None  0.3s
beginning cleanup
Test Summary: |Time
Cleanup       | None  1.5s
Test Summary:    | Pass  Total  Time
non-jl revisions |    4      4  2.0s
beginning signatures tests
┌ Warning: Replacing docs for `Base.CoreLogging.handle_message :: Union{}` in module `Base.CoreLogging`
└ @ Base.Docs docs/Docs.jl:243
┌ Warning: Replacing docs for `Base.CoreLogging.shouldlog :: Union{}` in module `Base.CoreLogging`
└ @ Base.Docs docs/Docs.jl:243
┌ Warning: Replacing docs for `Base.CoreLogging.min_enabled_level :: Union{}` in module `Base.CoreLogging`
└ @ Base.Docs docs/Docs.jl:243
┌ Warning: Replacing docs for `Base.CoreLogging.catch_exceptions :: Tuple{Any}` in module `Base.CoreLogging`
└ @ Base.Docs docs/Docs.jl:243
┌ Warning: Replacing docs for `Base.CoreLogging.NullLogger :: Union{}` in module `Base.CoreLogging`
└ @ Base.Docs docs/Docs.jl:243
WARNING: redefinition of constant CoreLogging._min_enabled_level. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant CoreLogging.CURRENT_LOGSTATE. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant CoreLogging.closed_stream. This may fail, cause incorrect answers, or produce other errors.
 45.478598 seconds (69.56 M allocations: 4.841 GiB, 5.17% gc time, 5.68% compilation time)
Base signatures: Test Failed at /home/rag/.julia/dev/Revise/test/sigtest.jl:133
  Expression: length(realfailed) < 60
   Evaluated: 354 < 60

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Test/src/Test.jl:679 [inlined]
 [2] top-level scope
   @ ~/.julia/dev/Revise/test/sigtest.jl:515
 [3] include(fname::String)
   @ Main ./sysimg.jl:38
 [4] macro expansion
   @ ~/.julia/dev/Revise/test/runtests.jl:3779 [inlined]
 [5] macro expansion
   @ ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Test/src/Test.jl:1598 [inlined]
 [6] top-level scope
   @ ~/.julia/dev/Revise/test/runtests.jl:3777
Test Summary:         | Pass  Fail  Total   Time
Base signatures       |    1     1      2  47.9s
  :lambda expressions |    1            1   0.1s
ERROR: LoadError: Some tests did not pass: 1 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/rag/.julia/dev/Revise/test/runtests.jl:3776
ERROR: Package Revise errored during testing
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/Types.jl:70
 [2] test(ctx::Pkg.Types.Context, pkgs::Vector{…}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
   @ Pkg.Operations ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/Operations.jl:2030
 [3] test
   @ ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/Operations.jl:1911 [inlined]
 [4] test(ctx::Pkg.Types.Context, pkgs::Vector{…}; coverage::Bool, test_fn::Nothing, julia_args::Cmd, test_args::Vector{…}, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool, kwargs::@Kwargs{…})
   @ Pkg.API ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/API.jl:444
 [5] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Pkg.UnstableIO, kwargs::@Kwargs{test_args::Vector{String}})
   @ Pkg.API ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/API.jl:159
 [6] test(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::@Kwargs{…})
   @ Pkg.API ~/.julia/juliaup/julia-nightly/share/julia/stdlib/v1.11/Pkg/src/API.jl:174
 [7] top-level scope
   @ REPL[7]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by timholy** [comment]
Thanks! I'm noticing some highly suspicious lines in the output:
```
WARNING: redefinition of constant Base.env_dict. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant Base.env_lock. This may fail, cause incorrect answers, or produce other errors.
```
Excluding `endswith(file, "env.jl")` from [sigtest.jl]() indeed prevents the segfault.

If I take these lines:
https://github.com/JuliaLang/julia/blob/567754cb96d55559c31ae2c24b686e22226e8d45/base/env.jl#L4C5-L21C8

and then do this:
```
julia> using JuliaInterpreter, LoweredCodeUtils

julia> ex = quote
       const ERROR_ENVVAR_NOT_FOUND = UInt32(203)

           const env_dict = IdDict{String, Vector{Cwchar_t}}()
           const env_lock = ReentrantLock()

           function memoized_env_lookup(str::AbstractString)
               # Windows environment variables have a different format from Linux / MacOS, and previously
               # incurred allocations because we had to convert a String to a Vector{Cwchar_t} each time
               # an environment variable was looked up. This function memoizes that lookup process, storing
               # the String => Vector{Cwchar_t} pairs in env_dict
               var = get(env_dict, str, nothing)
               if isnothing(var)
                   var = @lock env_lock begin
                       env_dict[str] = cwstring(str)
                   end
               end
               var
           end
       end;

julia> module ModSelective end;

julia> frame = Frame(ModSelective, ex);

julia> src = frame.framecode.src;

julia> edges = CodeEdges(src);

julia> isrequired = falses(length(src.code));

julia> isrequired[20] = true;   # visual inspection shows that statement 20 is the one that defines the method

julia> LoweredCodeUtils.lines_required!(isrequired, src, edges);

julia> LoweredCodeUtils.print_with_code(stdout, src, isrequired)
```
then the output (as a screenshot to show color) is
![image](https://github.com/timholy/Revise.jl/assets/1525481/5af23fb7-6aad-4aa7-8735-41f4d9355f9d)

You can see that statements 10 and 13 assign `env_dict` and `env_lock`, respectively, and are marked `t` indicating that they are scheduled for evaluation. That's because `edges` indicates these are used by statement 20:
```julia
CodeEdges:
  String: assigned on ∅, depends on ∅, and used by [6]
  env_lock: assigned on [13], depends on [12], and used by [11, 20]
  Vector: assigned on ∅, depends on ∅, and used by [7]
  env_dict: assigned on [10], depends on [9], and used by [4, 20]
  nothing: assigned on ∅, depends on ∅, and used by [20]
  IdDict: assigned on ∅, depends on ∅, and used by [5]
  AbstractString: assigned on ∅, depends on ∅, and used by [17]
  memoized_env_lookup: assigned on [15, 20], depends on [15, 19, 20], and used by [14, 15, 16, 20, 21]
  ERROR_ENVVAR_NOT_FOUND: assigned on [3], depends on [2], and used by [1]
  Cwchar_t: assigned on ∅, depends on ∅, and used by [7]
  statement  1 depends on [3] and is used by ∅
  statement  2 depends on ∅ and is used by [3]
  statement  3 depends on [2] and is used by [1]
  statement  4 depends on [10] and is used by ∅
  statement  5 depends on ∅ and is used by [8]
  statement  6 depends on ∅ and is used by [8]
  statement  7 depends on ∅ and is used by [8]
  statement  8 depends on [5, 6, 7] and is used by [9]
  statement  9 depends on [8] and is used by [10]
  statement 10 depends on [9] and is used by [4, 20]
  statement 11 depends on [13] and is used by ∅
  statement 12 depends on ∅ and is used by [13]
  statement 13 depends on [12] and is used by [11, 20]
  statement 14 depends on ∅ and is used by ∅
  statement 15 depends on [20] and is used by ∅
  statement 16 depends on [15, 20] and is used by [17]
  statement 17 depends on [16] and is used by [19]
  statement 18 depends on ∅ and is used by [19]
  statement 19 depends on [17, 18] and is used by [15, 20]
  statement 20 depends on [15, 19] and is used by ∅
  statement 21 depends on [15, 20] and is used by ∅
```
In this case, it looks like LoweredCodeUtils is too smart for its own good, discovering the dependency and thinking that it has to be evaluated.

**Comment by timholy** [comment]
@Zentrik, nightly isn't very well supported yet. Need to get LoweredCodeUtils passing its tests first.

**Comment by Zentrik** [comment]
Please disregard what I sent anyways, I forgot I was on linux.

**Comment by timholy** [comment]
Possible fix (thinking out loud...): if Revise is in `mode=:sigs`, any time we see an assignment to a `const`, we first check if that `const` is already defined. If so, we `empty!(edges.byname[varname].succs)` before calling `lines_required!`. We might also want to include the `.assigned` line(s) in `norequire`?

**Comment by timholy** [comment]
Fix in #791. Thanks all!
